File: DataAlgo_forPy/Trees/tree_level_order.rb
# Given a binary tree of integers, print it in level order.

# The output will contain space between the numbers in the same level,
# and new line between different levels.
# For example, if the tree is:

#          1
#        /   \
#       2      3
#     /       /  \
#    4       5    6

#    output:
#    1
#    2 3
#    4 5 6

class Node
  attr_accessor :left, :right
  attr_reader :value

  def initialize(value, right = nil, left = nil)
    @value = value
    @right = right
    @left = left
  end
end

def level_order_print(tree)
  return nil if tree.nil?

  nodes = [tree]

  current_count = 1
  next_count = 0

  until nodes.empty?
    current_node = nodes.shift
    current_count -= 1

    print current_node.value

    if current_node.left
      nodes << current_node.left
      next_count += 1
    end

    if current_node.right
      nodes << current_node.right
      next_count += 1
    end

    if current_count.zero?
      p '' # print new line
      current_count, next_count = next_count, current_count
    end
  end

end


root = Node.new(1)
left = Node.new(2)
right = Node.new(3)
sub_left = Node.new(4)
sub_right_left = Node.new(5)
sub_right_right = Node.new(6)

root.left = left
root.right = right
left.left = sub_left
right.left = sub_right_left
right.right = sub_right_right

level_order_print(root)

File: DataAlgo_forPy/Trees/tree_level_order.js
// Given a binary tree of integers, print it in level order.

// The output will contain space between the numbers in the same level,
// and new line between different levels.
// For example, if the tree is:

//          1
//        /   \
//       2      3
//     /       /  \
//    4       5    6

//    output:
//    1
//    2 3
//    4 5 6

  class Node {
    constructor(value, left = null, right = null) {
      this.value = value
      this.left = left
      this.right = right
    }
  }

const tree_level_order = (tree) => {
  if (tree === null) { return null }

  const nodes = [tree]

  let current_count = 1
  let next_count = 0

  while (nodes.length > 0) {
    let current_node = nodes.shift()
    current_count--

    console.log(current_node.value)

    if (current_node.left) {
      nodes.push(current_node.left)
      next_count++
    }

    if (current_node.right) {
      nodes.push(current_node.right)
      next_count++
    }

    if (current_count === 0) {
      console.log("\n")
      current_count = next_count
      next_count = 0
    }
  }

}


const root = new Node(1)
const left = new Node(2)
const right = new Node(3)
const sub_left = new Node(4)
const sub_right_left = new Node(5)
const sub_right_right = new Node(6)

root.left = left
root.right = right
left.left = sub_left
right.left = sub_right_left
right.right = sub_right_right

console.log(tree_level_order(root));
